### ISSUE TITLE
Possible to add a search filter that allows free-form entry?

### ISSUE THREAD

**Issue by artdirk** [adobe/asset-share-commons#265]
One of the teams planning to deploy an asset share has an asset metadata attribute that's a simple free-form entry field, and here's how it's represented as a search filter within the AEM author instance:

![image](https://user-images.githubusercontent.com/33265694/45187953-5f588a80-b1e7-11e8-94a2-2f3bf8267d6f.png)

Would it be possible to add a new Asset Share Commons search filter component to support this use case, provided something does not already exist (I confirmed the available search filters and none of them appeared to handle this use case)?

**Comment by davidjgonzalez** [comment]
@artdirk for sure! this is a good idea for sure -- i can totally appreciate where typing in wil be faster than fumbling w selection controls (ie picking a year from a list fo 50 years).

What sort of "matching" are you envisioning? Equals? Contains? Starts with?

**Comment by artdirk** [comment]
@davidjgonzalez, you bring up a very good point!

While the use case for the "Year" filter I mention would be an "equals" match, I can envision circumstances where the business may want to have the filter configured for one of the other types of matches you mention.

I'm assuming it would be possible to add a configuration that lets the author specify the type of match?

**Comment by davidjgonzalez** [comment]
Assuming we can construct a query builder predicate that can do it, and it performs in a performant manner - then yes! but we'd have to make sure we enumerate what these options are - we could do equals as a v1 unless there are other operations.

Also, this might be overkill, but should the input fields have any restrictions? textfield vs numberfield? or ability to specify html5 validations - wdyt? maybe validation would be sufficient?

would we want to prevent someone from entering text into the year field? Or require it to be 4 numners? `\d{4}` .. for example.

**Comment by davidjgonzalez** [comment]
@artdirk thoughts?

**Comment by artdirk** [comment]
@davidjgonzalez, per your comments I think allowing the author to specify the following for this new search filter component would be very helpful:

- Field type (text or number)
- Match type (equals, contains or starts with - default will be "equals" if not specified)
- Number of characters (allows the author to enter a number - default will be no limit if not specified)

How does this sound to you? Thanks!

**Comment by davidjgonzalez** [comment]
FYI - still working on this in some spare time; turns out to do this right its a lot more complicated than i initially thought :) 

Getting closer.... (feature/265 on my fork if youre interested in poking around)

**Comment by artdirk** [comment]
Thanks for the updates @davidjgonzalez!

As an FYI, I discovered that the Asset Share Commons search field will find assets that have the Year metadata property set.

For example, I set the Year metadata property to "2012" for an asset and ensured that "2012" was not included in the asset's name, and it was retrieved when I searched for "2012".

This somewhat mitigates the immediate need for this filter component for our business, but it would still be very valuable to have it available for less intuitive metadata properties.
